pkg/healthcheck/healthcheck_output.go

```diff
@@ -114,10 +142,109 @@ func PrintChecksResult(wout io.Writer, output string, success bool, warning bool
+func GetCLIChecks(pathEnv string, glob Glob, exec utilsexec.Interface) CLIChecks {
+	checks := CLIChecks{}
+
+	for _, dir := range filepath.SplitList(pathEnv) {
+		matches, err := glob(filepath.Join(dir, "linkerd-*"))
+		if err != nil {
+			continue
+		}
+
+		for _, match := range matches {
+			path, err := exec.LookPath(match)
+			if err != nil {
+				continue
+			}
+
+			cmd := exec.Command(path, "check-cli")
+			var stdout, stderr bytes.Buffer
+			cmd.SetStdout(&stdout)
+			cmd.SetStderr(&stderr)
+			err = cmd.Run()
+			if err != nil {
+				continue
+			}
+
+			checkCLIOutput, err := parseJSONCheckCLIOutput(stdout.Bytes())
+			if err != nil {
+				continue
+			}
+
+			_, filename := filepath.Split(path)
+			if checkCLIOutput.Name != filename {
+				// output of check-cli must match the executable name
+				// i.e. linkerd-foo is allowed, linkerd-foo-v0.XX.X is not
+				continue
+			}
+
+			if _, ok := checks[checkCLIOutput]; ok {
+				// this check already found earlier in the path, skip it
+				continue
+			}
+
+			checks[checkCLIOutput] = path
+		}
+	}
+
+	return checks
+}
+
+// RunExtensionsChecks runs both CLI and cluster extension checks.
+func RunExtensionsChecks(
+	wout io.Writer, werr io.Writer, cliChecks CLIChecks, extensions []string, flags []string, output string,
+) (bool, bool) {
+	cliSuccess, cliWarning := runCLIChecks(wout, werr, cliChecks, flags, output)
+	extSuccess, extWarning := runExtensionsChecks(wout, werr, extensions, flags, output)
+
+	return cliSuccess && extSuccess, cliWarning || extWarning
+}
+
+func runCLIChecks(wout io.Writer, werr io.Writer, cliChecks CLIChecks, flags []string, output string) (bool, bool) {
+	if len(cliChecks) == 0 {
+		return true, false
+	}
+
+	spin := spinner.New(spinner.CharSets[9], 100*time.Millisecond)
+	spin.Writer = wout
+
+	success := true
+	warning := false
+	for c, path := range cliChecks {
+		args := append([]string{"check"}, flags...)
+		checkResults, checkSuccess := runCheck(path, args, c.suffix(), CategoryID(c.Name), spin)
+		if !checkSuccess {
+			success = false
+		}
+		results := CheckResults{
+			Results: checkResults,
+		}
+
+		extensionSuccess, extensionWarning := RunChecks(wout, werr, results, output)
+		if !extensionSuccess {
+			success = false
+		}
+		if extensionWarning {
+			warning = true
+		}
+	}
+
+	return success, warning
+}
+
+// runExtensionsChecks runs checks for each extension name passed into the `extensions` parameter
 // and handles formatting the output for each extension's check. This function also handles
 // finding the extension in the user's path and runs it.
-func RunExtensionsChecks(wout io.Writer, werr io.Writer, extensions []string, flags []string, output string) (bool, bool) {
+func runExtensionsChecks(wout io.Writer, werr io.Writer, extensions []string, flags []string, output string) (bool, bool) {
+	if len(extensions) == 0 {
+		return true, false
+	}
+
 	if output == TableOutput {
 		PrintChecksHeader(wout, extensionsHeader)
```

Comment:
A few practical consequences of this to keep in mind:
* We can remove the concept of the `check-cli` extension command from this PR and treat every executable on the PATH named `linkerd-*` as an extension
  * Although we probably want to filter out `linkerd-edge-*` and `linkerd-stable-*`.
* We can merge the list of extensions discovered from the cluster and the list of extensions discovered from the PATH and run checks on all of them in a uniform way
* This change introduces a behavioral change where checks which are present locally but not installed on the cluster will now be run.  I think this is a good change, but we'll want to make sure we appropriately call it out in e.g. upgrade notes.
* The viz, jaeger, and multicluster extensions are bundled within the Linkerd CLI rather than being separate extension executables.  Therefore, I think we should special case these checks to only run when installed on the cluster (because it's not possible to exclude these from your PATH).

Reply:
The check-cli command feels like a valuable addition. It establishes an API (of sorts) to ensure that the linkerd-* item found on the path is meant for this. I think that should be retained.

I'm concerned about changing to defaulting behavior to run the viz, jaeger, and multicluster extension even if they aren't present on the cluster. That seems riskier for existing users (say ones that have incorporated check into CI flows) rather than leaving the behavior as-is while adding support to incorporate local path-based extensions.